### DISCUSSION TITLE
How to implement older protocol versions

### DISCUSSION THREAD

**Discussion by horshees** [semuconsulting/pyubx2#129] [Q&A] [answered]
Hey,
First of all, thanks for all the work you done on this project, I'm using this and PyGPSClient a lot and it's great to work with.
I'm currently using M8 generation receivers and was looking into the TIM-TP message. I wanted to read the "TpNotLocked" value from the flags section. As I could not find it while it is in the protocol description (UBX-13003221 - R28) I locally added it. I was looking into contributing this as a change but double checked a more recent protocol description(UBX-22008968 - R01) and saw the flag has been removed.
Is there, or should there be a way to incorporate older versions of the protocol to include these situations?

**Comment by semuadmin** [comment, marked as answer]
Hi @horshees,

There's no major issue adding additional bit flag definitions to an existing bitfield e.g. ...

```
    "TIM-TP": {
        "towMS": U4,
        "towSubMS": [U4, 2**-32],
        "qErr": I4,
        "week": U2,
        "flags": (
            X1,
            {
                "timeBase": U1,
                "utc": U1,
                "raim": U2,
                "qErrInvalid": U1,
                "TpNotLocked": U1,    # deprecated bit flag
            },
        ),
        "refinfo": (
            X1,
            {
                "timeRefGnss": U4,
                "utcStandard": U4,
            },
        ),
    },
```
... but in the absence of an explicit `version` attribute or other way to distinguish between different generations of a given message definition (e.g. payload length), you would have to apply the change globally and leave users to exercise their own discretion as to whether this particular bit flag was relevant to their particular device and firmware version.

`pyubx2.UBXMessage` does already incorporate 'version detection' for a number of legacy message definitions (including CFG-NMEA, NAV-RELPOSNED, RXM-PMP, RXM-PMREQ, RXM-RLM & TIM-VCOCAL), typically based on `version` or payload length, but this has to be done on a message-by-message basis as historically there's been very little consistency in UBX's approach to versioning payload definitions. Most new message definitions include an explicit `version` attribute but this is not true of older generations.

FYI This is actually why `pyubx2.UBXMessage` incorporates a `parsebitfield` keyword argument - to give users the option to parse the bitfield as a single hex value rather than a series of individual bit flags, so you're not tied to a specific protocol generation.

Hope this helps - any further queries, by all means get back to me.

**Comment by semuadmin** [comment]
FYI I've uploaded a [commit](https://github.com/semuconsulting/pyubx2/commit/dfbdef004b015cd03e49e9c76471aa79fdf7dc3e) which includes an amended TIM-TP payload definition, but if you want to submit a PR yourself for the change, be my guest. Contributions always welcome.
